HARK/ConsumptionSaving/ConsAggShockModel.py

```diff
@@ -550,11 +550,326 @@ def getMrkvNow(self):  # This function exists to be overwritten in StickyE model
+                              [GB_unemp_permute, GG_unemp_permute]]
+        self.emp_permute =   [[BB_emp_permute, BG_emp_permute],
+                              [GB_emp_permute, GG_emp_permute]]
+        
+    def reset(self):
+        self.initializeSim()
+        
+    def marketAction(self):
+        self.simulate(1)
+    
+    def initializeSim(self):
+        self.MrkvNow = self.MrkvInit
+        self.MrkvPrev = self.MrkvInit
+        AgentType.initializeSim(self)
+        self.EmpNow = self.EmpNow.astype(bool)
+        self.makeEmpIdxArrays()
+        
+    def simBirth(self,which):
+        '''
+        Create newborn agents with randomly drawn employment states.  This will
+        only ever be called by initializeSim() at the start of a new simulation
+        history, as the Krusell-Smith model does not have death and replacement.
+        The simDeath() method does not exist, as AgentType's default of "no death"
+        is the correct behavior for the model.
+        '''
+        N = np.sum(which)
+        if N == 0:
+            return
+        
+        if self.MrkvNow == 0:
+            unemp_N = int(np.round(self.UrateB*N))
+            emp_N = self.AgentCount - unemp_N
+        elif self.MrkvNow == 1:
+            unemp_N = int(np.round(self.UrateG*N))
+            emp_N = self.AgentCount - unemp_N
+        else:
+            assert False, 'Illegal macroeconomic state: MrkvNow must be 0 or 1'
+        EmpNew = np.concatenate([np.zeros(unemp_N, dtype=bool), np.ones(emp_N, dtype=bool)])
+        self.EmpNow[which] = self.RNG.permutation(EmpNew)
+        self.aNow[which] = self.kInit
+    
+    def getShocks(self):
+        '''
+        Get new idiosyncratic employment states based on the macroeconomic state.
+        '''
+        # Get boolean arrays for current employment states
+        employed = self.EmpNow.copy()
+        unemployed = np.logical_not(employed)
+        
+        # Transition some agents between unemployment and employment
+        emp_permute = self.emp_permute[self.MrkvPrev][self.MrkvNow]
+        unemp_permute = self.unemp_permute[self.MrkvPrev][self.MrkvNow]
+        EmpNow = self.EmpNow
+        EmpNow[employed] = self.RNG.permutation(emp_permute)
+        EmpNow[unemployed] = self.RNG.permutation(unemp_permute)
+
+    def getStates(self):
+        '''
+        Get each agent's idiosyncratic state, their household market resources.
+        '''
+        self.mNow = self.Rnow*self.aNow + self.Wnow*self.LbrInd*self.EmpNow
+
+    def getControls(self):
+        '''
+        Get each agent's consumption given their current state.'
+        '''
+        Mnow = self.Mnow*np.ones(self.AgentCount)
+        cNow = np.zeros(self.AgentCount)
+        
+        # This can be written more concisely, but is intentionally explicit for
+        # the sake of making it very easy to read
+        if self.MrkvNow == 0: # Bad macroeconomic conditions
+            these = np.logical_not(self.EmpNow) # Unemployed agents
```

Comment:
maybe name this variable `unemployed_agents` instead of `these`. 
Save you having to explain it in comment :)

Reply:
I suppose, but this is used throughout HARK for binary indexing arrays.

Reply:
fair enough. just a suggestion.